cluster-autoscaler/cloudprovider/gce/gce_cloud_provider.go

```diff
@@ -0,0 +1,261 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gce
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"k8s.io/contrib/cluster-autoscaler/cloudprovider"
+	kube_api "k8s.io/kubernetes/pkg/api"
+)
+
+// NodeProvider is a function that provides a list of nodes.
+type NodeProvider func() ([]*kube_api.Node, error)
+
+// GceCloudProvider implements CloudProvider interface.
+type GceCloudProvider struct {
+	gceManager   *GceManager
+	migs         []*Mig
+	nodeProvider NodeProvider
+}
+
+// BuildGceCloudProvider build CloudProvider implementation for GCE.
+func BuildGceCloudProvider(gceManager *GceManager, nodeProvider NodeProvider) *GceCloudProvider {
+	return &GceCloudProvider{
+		gceManager:   gceManager,
+		migs:         make([]*Mig, 0),
+		nodeProvider: nodeProvider,
+	}
+}
+
+// AddNodeGroup adds node group defined in string spec. Format:
+// minNodes:maxNodes:migUrl
+func (gce *GceCloudProvider) AddNodeGroup(spec string) error {
+	mig, err := buildMig(spec, gce.gceManager)
+	if err != nil {
+		return err
+	}
+	nodes, err := gce.nodeProvider()
+	if err != nil {
+		return err
+	}
+	for _, node := range nodes {
+		if belongs, err := mig.Belongs(node); err == nil && belongs {
```

Comment:
As we discussed before we probably want to revisit the way how sample node is being chosen. Maybe add a TODO here to do it better?